lib/Dialect/FIRRTL/FIRRTLOps.cpp

```diff
@@ -891,30 +906,48 @@ Value MemOp::getPortNamed(StringAttr name) {
 //===----------------------------------------------------------------------===//
 
 static LogicalResult verifyConnectOp(ConnectOp connect) {
-  FIRRTLType destType =
-      connect.dest().getType().cast<FIRRTLType>().getPassiveType();
-  FIRRTLType srcType =
-      connect.src().getType().cast<FIRRTLType>().getPassiveType();
+  FIRRTLType destType = connect.dest().getType().cast<FIRRTLType>();
+  FIRRTLType srcType = connect.src().getType().cast<FIRRTLType>();
 
   // Analog types cannot be connected and must be attached.
   if (destType.isa<AnalogType>() || srcType.isa<AnalogType>())
     return connect.emitError("analog types may not be connected");
 
+  if (destType.getPassiveType().isa<BundleType>()) {
+    // For bulk connections, we need to make sure that the connection is
+    // unambiguous by making sure that both sides are not duplex types. TODO: we
+    // are not checking that the connections are recursively well formed when
+    // neither is a duplex type.
+    if (isDuplexValue(connect.dest()) && isDuplexValue(connect.src())) {
+      return connect.emitOpError()
+             << "Connection between source and destination "
+             << "failed, unclear which side is the LHS as both operands are "
+             << "duplex types. You can not use bulk connections here.";
+    }
+  } else {
+    // This is a mono-connection. Check that the LHS side is a sink or duplex.
+    // Since we can read from a either a passive or flip type, we don't need to
+    // check anything on the RHS.
+    if (destType.isPassive() && !isDuplexValue(connect.dest())) {
+      return connect.emitOpError("Connection destination must be a non-passive "
+                                 "type or a duplex value");
+    }
+  }
+
   // Destination and source types must be equivalent.
   if (!areTypesEquivalent(destType, srcType))
     return connect.emitError("type mismatch between destination ")
-           << destType << " and source " << srcType;
+           << destType.getPassiveType() << " and source "
```

Comment:
I added `getPassiveType()` here just so I didn't have to go change the error messages.  I personally think it would be a bit better if this printed the original types, not the passive version. Maybe a different PR?